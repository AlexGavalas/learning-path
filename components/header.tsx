import Link from 'next/link';
import styled from '@emotion/styled';
import { motion } from 'framer-motion';

import { Button } from './button';

const StyledHeader = styled.header`
    height: 5rem;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    border-block-end: 1px solid;
    width: 100%;
`;

const LinksContainer = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const HorizontalSeparator = styled.span`
    margin-inline: 0.5rem;
`;

const LoginContainer = styled.div``;

const Logo = styled(motion.h1)`
    cursor: pointer;
    padding: 0.5rem;
    border-radius: 0.5rem;
    background-color: #ffffff;
    font-size: clamp(1.5rem, 8vw, 2rem);
`;

export const Header = () => {
    return (
        <StyledHeader>
            <LinksContainer>
                <Link href="/" passHref>
                    <Logo
                        whileHover={{ backgroundColor: '#d7d7d7' }}
                        transition={{ duration: 0.5 }}
                    >
                        Learning Path
                    </Logo>
                </Link>
            </LinksContainer>
            {/* <LoginContainer>
				<Button>Login</Button>
			</LoginContainer> */}
        </StyledHeader>
    );
};
