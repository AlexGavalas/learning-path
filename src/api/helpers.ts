import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
