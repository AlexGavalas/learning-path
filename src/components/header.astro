---
import ThemeSwitch from '~components/theme-switch/theme-switch.astro';
import Logo from '~components/logo/logo.astro';
---

<header
    class="flex w-full items-center justify-between border-b border-solid border-b-light-border dark:border-b-dark-border"
>
    <a
        aria-label="Go to homepage"
        href="/"
        class="my-2 w-40 cursor-pointer rounded-lg p-2 hover:bg-gray-300 hover:no-underline dark:hover:bg-gray-800 xs:w-52"
    >
        <Logo />
    </a>
    <div>
        <ThemeSwitch />
    </div>
</header>
