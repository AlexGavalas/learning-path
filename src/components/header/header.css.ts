import { style } from '@vanilla-extract/css';

import { vars } from '../../styles/theme.css';

export const container = style({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottom: `1px solid ${vars.color.border}`,
});

export const logo = style({
    margin: `${vars.spacing.small} 0`,
    padding: vars.spacing.small,
    width: '30%',
    minWidth: '150px',
    borderRadius: '5px',
    ':hover': {
        backgroundColor: vars.color.backgroundHover,
    },
});
