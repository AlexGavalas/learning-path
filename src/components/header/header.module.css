.container {
    display: flex;
    align-items: center;
    justify-content: space-between;
    border-bottom: 1px solid var(--color-border);
}

.logo {
    margin: var(--spacing-sm) 0;
    padding: var(--spacing-sm);
    width: 30%;
    min-width: 150px;
    border-radius: var(--spacing-xs);

    &:hover {
        background-color: oklch(
            from var(--color-background) calc(l + 0.07) c h
        );
    }
}
