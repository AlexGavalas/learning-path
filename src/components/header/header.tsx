import { useEffect, useState } from 'react';

import { Button } from '~components/button';
import { DEFAULT_THEME, STORAGE_KEY, THEME } from '~constants';

const LIGHT_IMG_PROPS = {
    src: '/moon.svg',
    alt: 'Moon',
};

const DARK_IMG_PROPS = {
    src: '/sun.svg',
    alt: 'Sun',
};

export const Header = (): JSX.Element => {
    const [isMounted, setIsMounted] = useState(false);

    const [theme, setTheme] = useState(
        typeof localStorage !== 'undefined'
            ? localStorage.getItem(STORAGE_KEY.THEME) ?? DEFAULT_THEME
            : DEFAULT_THEME,
    );

    useEffect(() => {
        setIsMounted(true);
    }, []);

    const toggleTheme = (): void => {
        const newTheme = theme === THEME.LIGHT ? THEME.DARK : THEME.LIGHT;

        setTheme(newTheme);

        document.documentElement.classList.toggle(
            'dark',
            newTheme === THEME.DARK,
        );
        document.documentElement.style.setProperty('color-scheme', newTheme);

        localStorage.setItem(STORAGE_KEY.THEME, newTheme);
    };

    const image = theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;

    return (
        <header className="flex w-full items-center justify-between border-b border-solid border-b-light-border dark:border-b-dark-border">
            <a
                href="/"
                className="my-2 cursor-pointer rounded-lg p-2 text-2xl font-[700] text-black hover:bg-gray-300 hover:no-underline dark:text-white dark:hover:bg-gray-800 xs:text-3xl"
            >
                Learning Path
            </a>
            <div className="mr-3">
                {isMounted && (
                    <Button onClick={toggleTheme} variant="wrapper">
                        <img
                            src={image.src}
                            alt={image.alt}
                            width={20}
                            height={20}
                        />
                    </Button>
                )}
            </div>
        </header>
    );
};
