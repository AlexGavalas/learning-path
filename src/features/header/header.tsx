import { useTheme } from 'next-themes';
import Image from 'next/image';
import Link from 'next/link';
import { useEffect, useState } from 'react';

import { Button } from '~components/button';

export const Header = () => {
    const [mounted, setIsMounted] = useState(false);
    const { setTheme, resolvedTheme } = useTheme();

    useEffect(() => {
        setIsMounted(true);
    }, []);

    const toggleTheme = () => {
        setTheme(resolvedTheme === 'light' ? 'dark' : 'light');
    };

    const image =
        resolvedTheme === 'light'
            ? {
                  src: '/moon.svg',
                  alt: 'Moon',
              }
            : {
                  src: '/sun.svg',
                  alt: 'Sun',
              };

    return (
        <header className="flex w-full items-center justify-between border-0 border-b border-solid dark:border-b-white">
            <div className="flex">
                <Link
                    href="/"
                    className="my-2 cursor-pointer rounded-lg p-2 text-3xl font-[700] text-black hover:bg-gray-300 hover:no-underline dark:text-white dark:hover:bg-gray-800"
                >
                    Learning Path
                </Link>
            </div>
            {mounted && (
                <div className="mr-3">
                    <Button onClick={toggleTheme} variant="wrapper">
                        <Image
                            src={image.src}
                            alt={image.alt}
                            width={20}
                            height={20}
                        />
                    </Button>
                </div>
            )}
        </header>
    );
};
