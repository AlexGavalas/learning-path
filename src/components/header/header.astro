---
import ThemeSwitch from '~components/theme-switch/theme-switch.astro';
import Logo from '~components/logo/logo.astro';

import * as cn from './header.css';
---

<header class={cn.container}>
    <a aria-label="Go to homepage" id="logo" href="/" class={cn.logo}>
        <Logo />
    </a>
    <ThemeSwitch />
</header>

<script>
    const logo = document.getElementById('logo');

    logo?.addEventListener('mouseenter', () => {
        logo.querySelector('g')
            ?.getAnimations()
            .forEach((animation) => {
                animation.cancel();
                animation.play();
            });
    });
</script>
