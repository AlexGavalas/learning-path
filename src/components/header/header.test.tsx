import { axe } from 'jest-axe';

import { renderWithUser, screen } from '~test/helpers';

import { Header } from './header';

describe('<Header />', () => {
    it('renders', () => {
        const { container } = renderWithUser(<Header />);

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderWithUser(<Header />);

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });

    describe('when user toggles theme', () => {
        it('changes theme', async () => {
            const { user } = renderWithUser(<Header />);

            const button = screen.getByRole('button', { name: 'Moon' });

            await user.click(button);

            const buttonImage = screen.getByRole('img');

            expect(buttonImage).toHaveAttribute('alt', 'Sun');

            await user.click(button);

            expect(buttonImage).toHaveAttribute('alt', 'Moon');
        });
    });
});
