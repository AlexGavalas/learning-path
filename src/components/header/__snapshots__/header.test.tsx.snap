// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders 1`] = `
<div>
  <header
    class="flex w-full items-center justify-between border-b border-solid border-b-light-border dark:border-b-dark-border"
  >
    <a
      class="my-2 cursor-pointer rounded-lg p-2 text-2xl font-[700] text-black hover:bg-gray-300 hover:no-underline dark:text-white dark:hover:bg-gray-800 xs:text-3xl"
      href="/"
    >
      Learning Path
    </a>
    <div
      class="mr-3"
    >
      <button
        class="flex items-center rounded border-0 p-2 bg-transparent hover:bg-slate-300 dark:hover:bg-slate-600"
      >
        <img
          alt="Moon"
          height="20"
          src="/moon.svg"
          width="20"
        />
      </button>
    </div>
  </header>
</div>
`;
